### PR TITLE
feat!: option to replace `happy-dom` with `jsdom`

### PR DESCRIPTION
**Pull request by enkot**
Resolves https://github.com/danielroe/nuxt-vitest/issues/81

Adds possibility to replace `happy-dom` with `jsdom` by using additional `domEnvironment` option , e.g:
```ts
export default defineVitestConfig({
  test: {
    environment: 'nuxt',
    environmentOptions: {
      nuxt: {
        domEnvironment: 'jsdom'
      }
    }
  },
})

``` 

Note: It would be great if `vitest` has some API to extend environments specified in https://github.com/vitest-dev/vitest/tree/main/packages/vitest/src/integrations/env/.

### CODE REVIEW

**Comment by enkot** [comment]
@antfu @danielroe Need help with tests. Trying to understand why some tests that use `useRouter`/`useRoute` are not passing for `jsdom`:
https://github.com/danielroe/nuxt-vitest/blob/main/playground/tests/nuxt/index.spec.ts#L28
https://github.com/danielroe/nuxt-vitest/blob/main/playground/tests/nuxt/index.spec.ts#L37

**Comment by danielroe** [comment]
The latest version now supports custom options passed via `environmentOptions.nuxt`, and I'd recommend that we use this namespace for configuring use of `jsdom`/`happy-dom`.

**Comment by enkot** [comment]
Hey @oskarols, yes, tests are still failing for jsdom. I would not refuse help to investigate that :)

File: packages/vitest-environment-nuxt/package.json [hunk 1]
```diff
@@ -60,5 +59,10 @@
     "vitest": "^0.24.5 || ^0.26.0 || ^0.27.0 || ^0.28.0 || ^0.29.0 || ^0.30.0",
     "vue": "^3.2.45",
     "vue-router": "^4.0.0"
+  },
+  "devDependencies": {
```
Comment (oskarols) [inline packages/vitest-environment-nuxt/package.json:63]: This duplicates the `devDependencies` 👆 

I think this might have caused some of the changes in the `pnpm-lock` file.

File: packages/vitest-environment-nuxt/src/index.ts [hunk 2]
```diff
@@ -61,6 +59,7 @@ export default <Environment>{
     }
 
     win.$fetch = createFetch({ fetch: win.fetch, Headers: win.Headers as any })
+    win.scrollTo = () => {}
```
Comment (oskarols) [inline packages/vitest-environment-nuxt/src/index.ts:62]: Why is this needed? 😄

File: packages/vitest-environment-nuxt/src/env/jsdom.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { importModule } from 'local-pkg'
+import { EnvironmentNuxt } from '../types'
+
+export default <EnvironmentNuxt> async function (global, { jsdom = {} }) {
```
Comment (oskarols) [inline packages/vitest-environment-nuxt/src/env/jsdom.ts:4]: Is this a copy from the JSDOM env from Vitest? If so, couldn't we just import and use this from Vitest?

**Comment by oskarols** [comment]
@enkot 👋  I've dug into the failing unit tests. I fixed two of them in #232 since them failing had little to do with this PR, but rather was due to underlying issues.

The "can mock fetch requests" test is failing since JSDOM, unlike Happy-dom, does not have its own browser `fetch` implementation. The solution to this is to add a fetch polyfill. `cross-fetch` seems to be a sensible choice.

E.g. we add this package as a dependency and put something like this somewhere in the JSDOM env file (or conditionally do it in `packages/vitest-environment-nuxt/src/index.ts`):

```ts
  // @ts-ignore missing types, but we do not need these
  await import('cross-fetch/polyfill')
```

File: packages/vitest-environment-nuxt/src/env/jsdom.ts (hunk 3 above)
Comment (enkot) [reply packages/vitest-environment-nuxt/src/env/jsdom.ts:4]: We can't use it directly from Vitest, because we need to return window object here.

**Comment by oskarols** [comment]
@enkot My fixes have now been merged. Please rebase against `main` and some of the tests in this PR will stop failing 😄

File: packages/vitest-environment-nuxt/src/index.ts (hunk 2 above)
Comment (enkot) [reply packages/vitest-environment-nuxt/src/index.ts:62]: JSDom doesn't have `scrollTo` implemented.

**Comment by enkot** [comment]
The tests are passed 🥳 Thanks @oskarols!
But besides this we have some problems with TS:
```sh
tests/nuxt/auto-import-mock.spec.ts:2:10 - error TS2305: Module '"vitest-environment-nuxt/utils"' has no exported member 'mockNuxtImport'.
```
Which is not related to this PR.
сс @danielroe

**Comment by danielroe** [comment]
@enkot I don't see that error. Have you run `pnpm dev:prepare`?

**Comment by enkot** [comment]
> @enkot I don't see that error. Have you run `pnpm dev:prepare`?

Completely forgot about it 🙈. Yep, no errors now :)
